### PR TITLE
Support 'package.json' not in package root

### LINKED ISSUE #18872: Typescript 2.5.3 break date-fns types
**TypeScript Version:**  2.5.3

**Code**

`npm i date-fns` first

```ts
import * as format from 'date-fns/format'
```

**Expected behavior:**

no error

**Actual behavior:**

```
Could not find a declaration file for module 'date-fns/format'. '/path/to/my/project/node_modules/date-fns/format/index.js' implicitly has an 'any' type.
```

Notes: it works with typescript v2.5.2, when I update typescript to v2.5.3, it broke.

The types: https://github.com/date-fns/date-fns/blob/master/typings.d.ts

Original issue: https://github.com/date-fns/date-fns/issues/572

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #18872, #18988, #19129

### CODE REVIEW

File: tests/cases/compiler/moduleResolution_packageJson_notAtPackageRoot.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+{ "types": "types.d.ts" }
+
+// @Filename: /node_modules/foo/package.json
+{}
+
+// @Filename: /node_modules/foo/bar/types.d.ts
+export const x: number;
+
+// @Filename: /a.ts
+import { x } from "foo/bar";
```
Comment (sheetalkamat) [inline tests/cases/compiler/moduleResolution_packageJson_notAtPackageRoot.ts]: What about handling/testing ```foo/@bar```, ```@foo/bar```

Comment (andy-ms) [reply tests/cases/compiler/moduleResolution_packageJson_notAtPackageRoot.ts]: `@foo/bar` tested in `moduleResolution_packageJson_scopedPackage`. Changed this test to use `foo/@bar`.

**Comment by andy-ms** [comment]
@sheetalkamat Good to go?

**Comment by sheetalkamat** [comment]
@andy-ms It would also be good to have tests for ```foo/bar```, ```foo/@bar```, ```@foo/bar``` especially when it doesnt find the package.json in first check.. How come there were no baselines that got updated because the first check failed/didnt happen?

**Comment by andy-ms** [comment]
@sheetalkamat The reason no baselines were updated was because we only do a second check if the first check wasn't already for the package root. We apparently didn't test resolution in package subdirectories before this.

**Pull request merged**